datafusion/core/src/physical_plan/aggregates/row_hash.rs

```diff
@@ -120,6 +132,56 @@ use super::AggregateExec;
 /// hash table).
 ///
 /// [`group_values`]: Self::group_values
+///
+/// # Spilling
+///
+/// The sizes of group values and accumulators can become large. Before that causes
+/// out of memory, this hash aggregator spills those data to local disk using Arrow
+/// IPC format. For every input [`RecordBatch`], the memory manager checks whether
+/// the new input size meets the memory configuration. If not, spilling happens, and
+/// later stream-merge sort the spilled data to read back. As the rows cannot be
+/// grouped between spilled data stored on disk, the read back merged data needs to
+/// be re-grouped again.
+///
+/// ```text
+/// Partial Aggregation [batch_size = 2] (max memory = 3 rows)
```

Comment:
For partial aggregation, when the memory is insufficient, I don't think we need to spill or merge sort; we could safely write the current hash table to output and wait for the final aggregate to do the rest.

Reply:
We could do that.
I think the trade-off is that the data size inflated by duplicated rows in terms of group values vs sorting cost. I chose to sort here because (1) the output data is compatible with non-spilling (2) chose data size benefit over sorting cost.

I would wait for other reviewers opinion on this.

Reply:
Other systems I have worked on have taken the approach described by @yjshen  which is to only spill the final aggregation and not spill the partial aggregates.  The rationale for not spilling the partial aggregate is that spilling is *very* expensive as it does disk IO and sorting and thus is likely orders of magnitude slower than hash partitioning more values (which is what happens if you dump less well aggregated batches).

Thus I would recommend simply not spilling at except for [`AggregateMode::Final`](https://docs.rs/datafusion/latest/datafusion/physical_plan/aggregates/enum.AggregateMode.html) or [`AggregateMode::FinalPartitioned`](https://docs.rs/datafusion/latest/datafusion/physical_plan/aggregates/enum.AggregateMode.html)

Spilling only on the Final phase might also simplify some of the accounting / schema below, as the input and the merge file input would always be the intermediate schema, I think

Reply:
Changed to spilling only on the Final phase, and output early on the partial phase.